p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -0,0 +1,351 @@
+		return &node.Link{Cid: right}, path[1:], nil
+	default:
+		return nil, nil, errors.New("invalid path for inner node")
+	}
+}
+
+func (n nmtNode) Tree(path string, depth int) []string {
+	if path != "" || depth == 0 {
+		return nil
+	}
+
+	return []string{
+		"0",
+		"1",
+	}
+}
+
+func (n nmtNode) ResolveLink(path []string) (*node.Link, []string, error) {
+	obj, rest, err := n.Resolve(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	lnk, ok := obj.(*node.Link)
+	if !ok {
+		return nil, nil, errors.New("was not a link")
+	}
+
+	return lnk, rest, nil
+}
+
+func (n nmtNode) Copy() node.Node {
+	panic("implement me")
+}
+
+func (n nmtNode) Links() []*node.Link {
+	leftCid, err := cidFromNamespacedSha256(n.l)
+	if err != nil {
+		panic(fmt.Errorf("malformed nmtNode, cidFromNamespacedSha256(): %w", err))
+	}
+	rightCid, err := cidFromNamespacedSha256(n.r)
+	if err != nil {
+		panic(fmt.Errorf("malformed nmtNode, cidFromNamespacedSha256(): %w", err))
+	}
+	return []*node.Link{{Cid: leftCid}, {Cid: rightCid}}
+}
+
+func (n nmtNode) Stat() (*node.NodeStat, error) {
+	return &node.NodeStat{}, nil
+}
+
+func (n nmtNode) Size() (uint64, error) {
+	return 0, nil
+}
+
+type nmtLeafNode struct {
+	cid  cid.Cid
+	Data []byte
+}
+
+func (l nmtLeafNode) RawData() []byte {
+	return append([]byte{nmt.LeafPrefix}, l.Data...)
+}
+
+func (l nmtLeafNode) Cid() cid.Cid {
+	return l.cid
+}
+
+func (l nmtLeafNode) String() string {
+	return fmt.Sprintf(`
+leaf {
+	hash: 		%x,
+	len(Data): 	%v
+}`, l.cid.Hash(), len(l.Data))
+}
+
+func (l nmtLeafNode) Loggable() map[string]interface{} {
+	return nil
+}
+
+func (l nmtLeafNode) Resolve(path []string) (interface{}, []string, error) {
+	if path[0] == "Data" {
+		// TODO: store the data separately
+		// currently nmtLeafNode{Data:} contains the actual data
+		// instead, there should be a link in the leaf to the actual data
+		return nil, nil, nil
+	} else {
+		return nil, nil, errors.New("invalid path for leaf node")
+	}
+}
+
+func (l nmtLeafNode) Tree(path string, depth int) []string {
+	if path != "" || depth == 0 {
+		return nil
+	}
+
+	return []string{
+		"Data",
+	}
+}
+
+func (l nmtLeafNode) ResolveLink(path []string) (*node.Link, []string, error) {
+	obj, rest, err := l.Resolve(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	lnk, ok := obj.(*node.Link)
+	if !ok {
+		return nil, nil, errors.New("was not a link")
+	}
+	return lnk, rest, nil
+}
+
+func (l nmtLeafNode) Copy() node.Node {
+	panic("implement me")
+}
+
+func (l nmtLeafNode) Links() []*node.Link {
+	return []*node.Link{{Cid: l.Cid()}}
+}
+
+func (l nmtLeafNode) Stat() (*node.NodeStat, error) {
+	return &node.NodeStat{}, nil
+}
+
+func (l nmtLeafNode) Size() (uint64, error) {
+	return 0, nil
+}
+
+func cidFromNamespacedSha256(namespacedHash []byte) (cid.Cid, error) {
+	if got, want := len(namespacedHash), 2*namespaceSize+sha256.Size; got != want {
+		return cid.Cid{}, fmt.Errorf("invalid namespaced hash lenght, got: %v, want: %v", got, want)
+	}
```

Comment:
There is also a problem here. Including x2 namespaces in CID does not make them work as you are probably expecting. Particularly, appended namespaces won't have any effect if transferred over IPFS(Bitswap), as all the nodes [form CIDs themselves](https://github.com/ipfs/go-bitswap/blob/master/message/message.go#L222) after receiving blocks. Also, they won't be found on DHT, and usage of such CIDs may cause errors in multiple unexpected places, as this not how they are supposed to be used.

The only way to extend the CID/MH with custom data is by the use of [Prefix](https://github.com/ipfs/go-cid/blob/master/cid.go#L556) which, unfortunately, has constant length and probably can't be used for your case.

Reply:
Although introducing a custom multihash hash with an encoder may help here

Reply:
Good point! There are some tests but I didn't yet test this plugin over the network. TBH, I was surprised that it actually was possible to add data to the Cid as needed (something I found out by accident when trying to do some benchmarks with a simple merkle tree here: https://github.com/lazyledger/ipld-plugin-experiments). 

The rationale behind including the namespaces in the CIds is we wanted the Cid to (ideally) match the raw NMT hashes. This is (almost) the case right now. I'll run some tests and think of a solution if necessary. 

I guess the options are: 
 - _introducing a custom multihash hash with an encoder_  (I need to look into this -not sure how that would look like yet)
 - forking bitswap to enable nodes to accept and compute those namespace-including CIds
 - ?

Reply:
> Although introducing a custom multihash hash with an encoder may help here

Doesn't this mean that pref.Sum would only accept sha256 / 32 bytes effectively anyways:
https://github.com/ipfs/go-cid/blob/e530276a7008f5973e7da6640ed305ecc5825d27/cid.go#L563-L577